source/jsonParsing.ml

```diff
@@ -64,3 +64,201 @@ let optional_list_member ~f name json =
+      match v with
+      | `Os -> obj [] k d
+      | `As -> arr [] k d
+      | (`Null | `Bool _ | `String _ | `Float _) as v ->
+          k { Node.location = Some location; value = v } d
+      | _ ->
+          raise
+            (JsonmException
+               {
+                 message = "Encountered unexpected token or element";
+                 location = Location.from_decoded_range (Jsonm.decoded_range d);
+               })
+    and arr vs k d =
+      match dec d with
+      | `Ae ->
+          k
+            {
+              Node.location = Some (Location.from_decoded_range (Jsonm.decoded_range d));
+              value = `List (List.rev vs);
+            }
+            d
+      | v -> value v (fun v -> arr (v :: vs) k) d
+    and obj ms k d =
+      match dec d with
+      | `Oe ->
+          k
+            {
+              Node.location = Some (Location.from_decoded_range (Jsonm.decoded_range d));
+              value = `Assoc (List.rev ms);
+            }
+            d
+      | `Name n -> value (dec d) (fun v -> obj ((n, v) :: ms) k) d
+      | _ ->
+          raise
+            (JsonmException
+               {
+                 message = "Encountered unexpected token or element";
+                 location = Location.from_decoded_range (Jsonm.decoded_range d);
+               })
+    in
+    let d = Jsonm.decoder (`String input) in
+    value (dec d) (fun v _ -> v) d
+
+
+  module Util = struct
+    open Core
+
+    let get_type_error expected json =
+      let get_message current = "Expected " ^ expected ^ ", got " ^ current in
+      let message =
+        match json.Node.value with
+        | `Assoc _ -> get_message "object"
+        | `Int _ -> get_message "integer"
+        | `Float _ -> get_message "float"
+        | `Null -> get_message "null"
+        | `String _ -> get_message "string"
+        | `Bool _ -> get_message "boolean"
+        | `List _ -> get_message "list"
+      in
+      TypeError { message; json }
+
+
+    let keys node =
+      match node.Node.value with
+      | `Assoc children -> List.map ~f:fst children
+      | _ -> []
+
+
+    (* Extract a child element with a particular key from a a json object. Raises TypeError for
+       other types. *)
+    let member key node =
+      match node.Node.value with
+      | `Assoc children -> (
+          match List.Assoc.find children key ~equal:String.equal with
+          | Some child -> child
+          | None -> null_node)
+      | _ -> raise (get_type_error "object" node)
+
+
+    let member_nullable key node =
+      match node.Node.value with
+      | `Assoc _ -> member key node
+      | _ -> null_node
+
+
+    let to_bool_option node =
+      match node.Node.value with
+      | `Bool b -> Some b
+      | _ -> None
+
+
+    let to_string node =
+      match node.Node.value with
+      | `String s -> s
+      | `Int i -> Format.sprintf "%d" i
+      | `Float f -> Format.sprintf "%f" f
```

Comment:
Is this what `Yojson` does too? I don't think you would want an implicit conversion to string when calling `to_string`. Same for `to_int` below.

Reply:
No, we could migrate to strictly what YoJSON does!